Finish osu! hitobject and timing point parsing (plus small fixes)

cl8n opened the pull request:
The way I've implemented this, there are some neat quirks that weren't featured in osu!stable:
- Full sample additions can apply to slider edges in the same way that they apply to any other object, and it's still 100% backwards-compatible
- Sample additions can be anywhere between 2~5 values, not limited to strictly 2 or 5
- Timing meters can be at decimal value (useful for things like 7/8 timing sections, as they can now be represented with 3.5/4)
- BPMs and slider velocities are limited to positive values (after parsing)

I'm not really sure how to do the new combos after spinners, because the only mode-specific parsing is done per-hitobject
- [X] Allow colour hax
- [X] Parse Slider edge hitsounds / additions
- [X] Parse HitObject additions
- [ ] Automatically apply a new combo to objects after spinners
- [X] Force spinners to never have a new combo
- [X] Play hit sample audio on hit
- [ ] Play slider edge hit sample audio on hit
- [ ] Play slider body hit sample audio on hit
- [X] Parse ControlPoint
- [X] Use ControlPoint.Sample to control HitObject.Sample defaulting
- [X] Throw ArgumentException when split strings have invalid lengths
- [X] Throw UnknownHitObjectException when the parser receives an invalid type
- [X] Throw NotImplementedException when the osu! parser receives a Hold object

cl8n committed:
Add support for colour hax

cl8n committed:
Parse slider edge hitsounds

peppy commented on osu.Game/Modes/Objects/HitObject.cs:
```diff
@@ -16,6 +16,7 @@ public abstract class HitObject
         public virtual double EndTime => StartTime;
 
         public bool NewCombo { get; set; }
+        public int ComboOffset { get; set; } = 0;
```
probably called this `ComboColourOffset`.

cl8n committed:
Standardize SampleBank

cl8n committed:
Complete SampleInfo and HitSampleInfo

cl8n committed:
Rename ComboOffset to ComboColourOffset

cl8n committed:
Add ControlPointAt helper function

cl8n committed:
Complete ControlPoints

cl8n committed:
Add KiaiType

cl8n committed:
Fix Beatmap helper functions

cl8n committed:
Fix HitObject and ControlPoint parser handlers

cl8n committed:
Add HitObjectParser helper function

cl8n committed:
Add and use UnknownHitObjectException

cl8n committed:
Parse samples for all HitObjects

cl8n committed:
Add files to project

cl8n committed:
Parse slider edge samples

cl8n committed:
Force spinners to never have a new combo

cl8n committed:
Fix style and formatting
- CRLF line endings
- Organize using statements (and remove unnecessary ones)
- Prefer properties over bare fields
- Standardize styling
- Newlines at the end of files

cl8n committed:
Clean up BeatLengthAt

cl8n committed:
Fix SampleType in TestDecodeHitObjects

cl8n committed:
Fix slider edge addition parsing

cl8n committed:
Remove osu! instantiations from osu!mania ruleset
This was causing problems with HitObject parsing, since osu!mania uses
type 128

cl8n committed:
Default SampleSet to Normal

UselessToucan commented on osu.Game.Modes.Osu/Objects/OsuHitObjectParser.cs:
```diff
@@ -4,39 +4,86 @@
+            if (split.Length < 5)
+                throw new ArgumentException("OsuHitObjects must have at least 5 values");
+
+            var type = (OsuHitObject.HitObjectType)Convert.ToInt32(split[3], NumberFormatInfo.InvariantInfo);
```
I guess it will work without `NumberFormatInfo.InvariantInfo`. It applies to other Convert calls here too.

cl8n replied:
are you sure? some of these values have decimals, so I think invariant culture is necessary (or did you just mean for integers?)

UselessToucan replied:
Yes, I mean integers. Not sure about `double`.

UselessToucan commented on osu.Game.Modes.Osu/Objects/OsuHitObjectParser.cs:
```diff
@@ -4,39 +4,86 @@
+                            s.RepeatCount = Convert.ToInt32(split[6], NumberFormatInfo.InvariantInfo);
+                            for (int i = 0; i <= s.RepeatCount; i++)
+                                edgeSamples.Add(ParseHitSample(section, null, null));
+                            goto case 10;
```
```
case 8:
    s.RepeatCount = Convert.ToInt32(split[6], NumberFormatInfo.InvariantInfo);
    for (int i = 0; i <= s.RepeatCount; i++)
    edgeSamples.Add(ParseHitSample(section, null, null));
case 10:
    s.Sample = ParseHitSample(section, split[4], null);
    break;
```

UselessToucan commented on osu.Game.Modes.Osu/Objects/OsuHitObjectParser.cs:
```diff
@@ -63,23 +110,15 @@ public override HitObject Parse(string text)
-                            (int)Convert.ToDouble(temp[1], CultureInfo.InvariantCulture)
+                            (int)Convert.ToDouble(temp[0], NumberFormatInfo.InvariantInfo),
+                            (int)Convert.ToDouble(temp[1], NumberFormatInfo.InvariantInfo)
                         );
```
I would prefer this to be `(int)Convert.ToDouble(temp[1], NumberFormatInfo.InvariantInfo));`.

UselessToucan commented on osu.Game/Modes/Objects/HitObjectParser.cs:
```diff
@@ -2,15 +2,68 @@
+            {
+                case 5:
+                    hitSample.File = string.IsNullOrWhiteSpace(split[4]) ? null : split[4];
+                    goto case 4;
```
Can be removed

cl8n replied:
Case fallthroughs don't work in C# apparently... I was surprised too

peppy replied:
rather than use `goto`, i usually either group the two cases together or create a second switch statement with the shared logic.

WorldSEnder replied:
How about this:
```
if(split.Length >= 5) {
    //... set file
}
if(split.Length >= 4) {
   //... set volume
}
// you get the idea
```

peppy replied:
oh, that's what this is for?

why not just
```
for (int i = 0; i < split.Length; i++)
{
    var value = split[i];
    switch (i)
    {
        /// use value...
    }
}
```

UselessToucan commented on osu.Game/Modes/Objects/HitObjectParser.cs:
```diff
@@ -2,15 +2,68 @@
+                    hitSample.Volume = Convert.ToInt32(split[3], NumberFormatInfo.InvariantInfo) == 0 ?
+                        section.Volume :
+                        Convert.ToInt32(split[3], NumberFormatInfo.InvariantInfo);
+                    goto case 3;
```
Can be removed

UselessToucan commented on osu.Game/Modes/Objects/HitObjectParser.cs:
```diff
@@ -2,15 +2,68 @@
+                    hitSample.Bank = (SampleBank)Convert.ToInt32(split[2], NumberFormatInfo.InvariantInfo) == SampleBank.None ?
+                        section.Bank :
+                        (SampleBank)Convert.ToInt32(split[2], NumberFormatInfo.InvariantInfo);
+                    goto case 2;
```
Can be removed

cl8n committed:
Fix HitSample parsing

cl8n committed:
Play all hit samples per object
